### PR TITLE
Initial implementation of the Bevy Remote Protocol (Adopted)

### PR DESCRIPTION
**Pull request by mweatherley**
# Objective

Adopted from #13563.

The goal is to implement the Bevy Remote Protocol over HTTP/JSON, allowing the ECS to be interacted with remotely.

## Solution

At a high level, there are really two separate things that have been undertaken here:
1. First, `RemotePlugin` has been created, which has the effect of embedding a [JSON-RPC](https://www.jsonrpc.org/specification) endpoint into a Bevy application.
2. Second, the [Bevy Remote Protocol verbs](https://gist.github.com/coreh/1baf6f255d7e86e4be29874d00137d1d#file-bevy-remote-protocol-md) (excluding `POLL`) have been implemented as remote methods for that JSON-RPC endpoint under a Bevy-exclusive namespace (e.g. `bevy/get`, `bevy/list`, etc.).

To avoid some repetition, here is the crate-level documentation, which explains the request/response structure, built-in-methods, and custom method configuration:
<details>
  <summary>Click to view crate-level docs</summary>

```rust
//! An implementation of the Bevy Remote Protocol over HTTP and JSON, to allow
//! for remote control of a Bevy app.
//!
//! Adding the [`RemotePlugin`] to your [`App`] causes Bevy to accept
//! connections over HTTP (by default, on port 15702) while your app is running.
//! These *remote clients* can inspect and alter the state of the
//! entity-component system. Clients are expected to `POST` JSON requests to the
//! root URL; see the `client` example for a trivial example of use.
//!
//! The Bevy Remote Protocol is based on the JSON-RPC 2.0 protocol.
//!
//! ## Request objects
//!
//! A typical client request might look like this:
//!
//! ```json
//! {
//!     "method": "bevy/get",
//!     "id": 0,
//!     "params": {
//!         "entity": 4294967298,
//!         "components": [
//!             "bevy_transform::components::transform::Transform"
//!         ]
//!     }
//! }
//! ```
//!
//! The `id` and `method` fields are required. The `param` field may be omitted
//! for certain methods:
//!
//! * `id` is arbitrary JSON data. The server completely ignores its contents,
//!   and the client may use it for any purpose. It will be copied via
//!   serialization and deserialization (so object property order, etc. can't be
//!   relied upon to be identical) and sent back to the client as part of the
//!   response.
//!
//! * `method` is a string that specifies one of the possible [`BrpRequest`]
//!   variants: `bevy/query`, `bevy/get`, `bevy/insert`, etc. It's case-sensitive.
//!
//! * `params` is parameter data specific to the request.
//!
//! For more information, see the documentation for [`BrpRequest`].
//! [`BrpRequest`] is serialized to JSON via `serde`, so [the `serde`
//! documentation] may be useful to clarify the correspondence between the Rust
//! structure and the JSON format.
//!
//! ## Response objects
//!
//! A response from the server to the client might look like this:
//!
//! ```json
//! {
//!     "jsonrpc": "2.0",
//!     "id": 0,
//!     "result": {
//!         "bevy_transform::components::transform::Transform": {
//!             "rotation": { "x": 0.0, "y": 0.0, "z": 0.0, "w": 1.0 },
//!             "scale": { "x": 1.0, "y": 1.0, "z": 1.0 },
//!             "translation": { "x": 0.0, "y": 0.5, "z": 0.0 }
//!         }
//!     }
//! }
//! ```
//!
//! The `id` field will always be present. The `result` field will be present if the
//! request was successful. Otherwise, an `error` field will replace it.
//!
//! * `id` is the arbitrary JSON data that was sent as part of the request. It
//!   will be identical to the `id` data sent during the request, modulo
//!   serialization and deserialization. If there's an error reading the `id` field,
//!   it will be `null`.
//!
//! * `result` will be present if the request succeeded and will contain the response
//!   specific to the request.
//!
//! * `error` will be present if the request failed and will contain an error object
//!   with more information about the cause of failure.
//!
//! ## Error objects
//!
//! An error object might look like this:
//!
//! ```json
//! {
//!     "code": -32602,
//!     "message": "Missing \"entity\" field"
//! }
//! ```
//!
//! The `code` and `message` fields will always be present. There may also be a `data` field.
//!
//! * `code` is an integer representing the kind of an error that happened. Error codes documented
//!   in the [`error_codes`] module.
//!
//! * `message` is a short, one-sentence human-readable description of the error.
//!
//! * `data` is an optional field of arbitrary type containing additional information about the error.
//!
//! ## Built-in methods
//!
//! The Bevy Remote Protocol includes a number of built-in methods for accessing and modifying data
//! in the ECS. Each of these methods uses the `bevy/` prefix, which is a namespace reserved for
//! BRP built-in methods.
//!
//! ### bevy/get
//!
//! Retrieve the values of one or more components from an entity.
//!
//! `params`:
//! - `entity`: The ID of the entity whose components will be fetched.
//! - `components`: An array of fully-qualified type names of components to fetch.
//!
//! `result`: A map associating each type name to its value on the requested entity.
//!
//! ### bevy/query
//!
//! Perform a query over components in the ECS, returning all matching entities and their associated
//! component values.
//!
//! All of the arrays that comprise this request are optional, and when they are not provided, they
//! will be treated as if they were empty.
//!
//! `params`:
//! `params`:
//! - `data`:
//!   - `components` (optional): An array of fully-qualified type names of components to fetch.
//!   - `option` (optional): An array of fully-qualified type names of components to fetch optionally.
//!   - `has` (optional): An array of fully-qualified type names of components whose presence will be
//!      reported as boolean values.
//! - `filter` (optional):
//!   - `with` (optional): An array of fully-qualified type names of components that must be present
//!     on entities in order for them to be included in results.
//!   - `without` (optional): An array of fully-qualified type names of components that must *not* be
//!     present on entities in order for them to be included in results.
//!
//! `result`: An array, each of which is an object containing:
//! - `entity`: The ID of a query-matching entity.
//! - `components`: A map associating each type name from `components`/`option` to its value on the matching
//!   entity if the component is present.
//! - `has`: A map associating each type name from `has` to a boolean value indicating whether or not the
//!   entity has that component. If `has` was empty or omitted, this key will be omitted in the response.
//!
//! ### bevy/spawn
//!
//! Create a new entity with the provided components and return the resulting entity ID.
//!
//! `params`:
//! - `components`: A map associating each component's fully-qualified type name with its value.
//!
//! `result`:
//! - `entity`: The ID of the newly spawned entity.
//!
//! ### bevy/destroy
//!
//! Despawn the entity with the given ID.
//!
//! `params`:
//! - `entity`: The ID of the entity to be despawned.
//!
//! `result`: null.
//!
//! ### bevy/remove
//!
//! Delete one or more components from an entity.
//!
//! `params`:
//! - `entity`: The ID of the entity whose components should be removed.
//! - `components`: An array of fully-qualified type names of components to be removed.
//!
//! `result`: null.
//!
//! ### bevy/insert
//!
//! Insert one or more components into an entity.
//!
//! `params`:
//! - `entity`: The ID of the entity to insert components into.
//! - `components`: A map associating each component's fully-qualified type name with its value.
//!
//! `result`: null.
//!
//! ### bevy/reparent
//!
//! Assign a new parent to one or more entities.
//!
//! `params`:
//! - `entities`: An array of entity IDs of entities that will be made children of the `parent`.
//! - `parent` (optional): The entity ID of the parent to which the child entities will be assigned.
//!   If excluded, the given entities will be removed from their parents.
//!
//! `result`: null.
//!
//! ### bevy/list
//!
//! List all registered components or all components present on an entity.
//!
//! When `params` is not provided, this lists all registered components. If `params` is provided,
//! this lists only those components present on the provided entity.
//!
//! `params` (optional):
//! - `entity`: The ID of the entity whose components will be listed.
//!
//! `result`: An array of fully-qualified type names of components.
//!
//! ## Custom methods
//!
//! In addition to the provided methods, the Bevy Remote Protocol can be extended to include custom
//! methods. This is primarily done during the initialization of [`RemotePlugin`], although the
//! methods may also be extended at runtime using the [`RemoteMethods`] resource.
//!
//! ### Example
//! ```ignore
//! fn main() {
//!     App::new()
//!         .add_plugins(DefaultPlugins)
//!         .add_plugins(
//!             // `default` adds all of the built-in methods, while `with_method` extends them
//!             RemotePlugin::default()
//!                 .with_method("super_user/cool_method".to_owned(), path::to::my::cool::handler)
//!                 // ... more methods can be added by chaining `with_method`
//!         )
//!         .add_systems(
//!             // ... standard application setup
//!         )
//!         .run();
//! }
//! ```
//!
//! The handler is expected to be a system-convertible function which takes optional JSON parameters
//! as input and returns a [`BrpResult`]. This means that it should have a type signature which looks
//! something like this:
//! ```
//! # use serde_json::Value;
//! # use bevy_ecs::prelude::{In, World};
//! # use bevy_remote::BrpResult;
//! fn handler(In(params): In<Option<Value>>, world: &mut World) -> BrpResult {
//!     todo!()
//! }
//! ```
//!
//! Arbitrary system parameters can be used in conjunction with the optional `Value` input. The
//! handler system will always run with exclusive `World` access.
//!
//! [the `serde` documentation]: https://serde.rs/
```

</details>

### Message lifecycle

At a high level, the lifecycle of client-server interactions is something like this:
1. The client sends one or more `BrpRequest`s. The deserialized version of that is just the Rust representation of a JSON-RPC request, and it looks like this:
```rust
pub struct BrpRequest {
    /// The action to be performed. Parsing is deferred for the sake of error reporting.
    pub method: Option<Value>,

    /// Arbitrary data that will be returned verbatim to the client as part of
    /// the response.
    pub id: Option<Value>,

    /// The parameters, specific to each method.
    ///
    /// These are passed as the first argument to the method handler.
    /// Sometimes params can be omitted.
    pub params: Option<Value>,
}
```
2. These requests are accumulated in a mailbox resource (small lie but close enough).
3. Each update, the mailbox is drained by a system `process_remote_requests`, where each request is processed according to its `method`, which has an associated handler. Each handler is a Bevy system that runs with exclusive world access and returns a result; e.g.:
```rust
pub fn process_remote_get_request(In(params): In<Option<Value>>, world: &World) -> BrpResult { // ... }
```
4. The result (or an error) is reported back to the client.

## Testing

This can be tested by using the `server` and `client` examples. The `client` example is not particularly exhaustive at the moment (it only creates barebones `bevy/query` requests) but is still informative. Other queries can be made using `curl` with the `server` example running. 

For example, to make a `bevy/list` request and list all registered components:
```bash
curl -X POST -d '{ "jsonrpc": "2.0", "id": 1, "method": "bevy/list" }' 127.0.0.1:15702 | jq .
```

---

## Future direction

There were a couple comments on BRP versioning while this was in draft. I agree that BRP versioning is a good idea, but I think that it requires some consensus on a couple fronts:
- First of all, what does the version actually mean? Is it a version for the protocol itself or for the `bevy/*` methods implemented using it? Both?
- Where does the version actually live? The most natural place is just where we have `"jsonrpc"` right now (at least if it's versioning the protocol itself), but this means we're not actually conforming to JSON-RPC any more (so, for example, any client library used to construct JSON-RPC requests would stop working). I'm not really against that, but it's at least a real decision.
- What do we actually do when we encounter mismatched versions? Adding handling for this would be actual scope creep instead of just a little add-on in my opinion. 

Another thing that would be nice is making the internal structure of the implementation less JSON-specific. Right now, for example, component values that will appear in server responses are quite eagerly converted to JSON `Value`s, which prevents disentangling the handler logic from the communication medium, but it can probably be done in principle and I imagine it would enable more code reuse (e.g. for custom method handlers) in addition to making the internals more readily usable for other formats.

### CODE REVIEW

**Review by NthTensor — COMMENTED** [review]
Thank you for picking this up!

One small thing, I think we want to add protocol versioning. In my opinion, we should start at 1.0.0 and follow semantic versioning spec for all verbs, networking types, and server handlers, just like it was a public rust API.

This will allow compatibility ranges between the editor and game apps. I would prefer we not move this in lock-step with the bevy release; it would be best if BRP was somewhat more stable than the engine.

File: Cargo.toml
Comment (mweatherley) [reply Cargo.toml:3253]: Weird. Maybe that was related to this? #14414

Comment (alice-i-cecile) [reply Cargo.toml:3253]: Reasonable change to make, but not in this PR.

**Comment by soqb** [comment]
> One small thing, I think we want to add protocol versioning. In my opinion, we should start at 1.0.0 and follow semver for all verbs, networking types, and server handlers, just like it was a public rust API. The client should include the protocol version on every request.

i agree protocol versioning is important, but i'm not sure that we should start at 1.0 considering Bevy is not yet 1.0, so i think 0.1.0 is probably a better initial version.

**Comment by MiniaczQ** [comment]
BRP is mostly intended for editor, can we keep this as "experimental" and start versioning when the editor integrates it?
There will be more changes coming and there isn't even a consumer

**Comment by NthTensor** [comment]
> can we keep this as "experimental"

yeah thats fine

File: crates/bevy_remote/src/lib.rs
Comment (MiniaczQ) [inline crates/bevy_remote/src/lib.rs:634]: Not in scope for this PR, but mixing "editor" logic with game logic will cause trouble in debugging, like during system stepping you won't be able to use BRP, because it's just a "game logic" system.
Is there a proposed solution for this?

Comment (NthTensor) [reply crates/bevy_remote/src/lib.rs:634]: Probably to special-case out the systems that make the editor 'embassy' tools like this work. This discussion might be better suited to discord.

File: examples/remote/client.rs
Comment (ChristopherBiscardi) [inline examples/remote/client.rs:47]: Is there a strong reason to have this example be this manual and not use something like ureq (or any other client). I feel like most people looking to make a request against the server aren't going to be looking for "how to use `http1::handshake::`", etc, and the async isn't really necessary either.

I put the relevant diff up in a PR for consideration: https://github.com/mweatherley/bevy/pull/2

**Comment by mweatherley** [comment]
I basically made it so that `bevy_remote` is neutered entirely on `wasm` targets to satisfy CI. Maybe we should investigate a relaxation of the CI requirement that the entire workspace compile on web assembly at some point (by way of contrast, the dependencies of `bevy_remote` don't just do nothing on `wasm`, they don't compile at all).

**Comment by alice-i-cecile** [comment]
I'm okay with this being neutered on web for now. @mweatherley, are you happy with the state of this now?

**Comment by mweatherley** [comment]
> I'm okay with this being neutered on web for now. @mweatherley, are you happy with the state of this now?

Yeah, I'm fine with it, I think :) I guess what I would say is that I don't personally have any blocking concerns about the state it's in right now, but I think various things (e.g. documentation) could be definitely improved.

File: crates/bevy_remote/src/builtin_methods.rs
Comment (alice-i-cecile) [inline crates/bevy_remote/src/builtin_methods.rs:31]: I think that these types would be clearer as `BrpGetRequest` and so on, to mirror `BrpGetResponse`.

File: crates/bevy_remote/src/builtin_methods.rs
Comment (alice-i-cecile) [reply crates/bevy_remote/src/builtin_methods.rs:41]: Seconding this request; I'd really like to newtype this in some way.

File: crates/bevy_remote/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_remote/src/lib.rs:522]: Debug derives everywhere please.

File: crates/bevy_remote/src/builtin_methods.rs
Comment (mweatherley) [reply crates/bevy_remote/src/builtin_methods.rs:31]: Yeah, I feel a bit divided on this; they are named like this because the `BrpRequest` type encompasses the entire request, and the params are what is method-specific (and corresponds to the contents of `params` in the request).

**Review by alice-i-cecile — APPROVED** [review]
I'm broadly happy with this design and the quality here. There are things I would prefer to clean-up (primarily the `Vec<String>` pattern), but I won't block on doing that in this PR. It's still wild to me that this *works*.

**Review by ChristopherBiscardi — APPROVED** [review]
I've run the code and it worked for me and I think its important to merge this even if the set of methods/return values/etc isn't perfect (because we can make informed changes after its merged). 

The architecture is very much not "widely used restish http server" (like axum), so the documentation for how to add methods and work on the server-side of things is correspondingly even more important.

and I think that we'll want to be mindful of feature increases in the future (does this grow a socket/websocket/etc interface in the future, is there user-providable middleware, can plugins add their own routes, etc) and if those features push us close enough towards "roughly axum shape" (hyper + tower) probably consider actually using it rather than re-implementing it accidentally.

File: crates/bevy_remote/src/lib.rs
Comment (mweatherley) [reply crates/bevy_remote/src/lib.rs:336]: Ah. It does do that; this was just an un-updated doc comment.

File: examples/remote/client.rs
Comment (torsteingrindvik) [inline examples/remote/client.rs:44]: Is this meant to show users specifically that "it's just JSON, so any language would work, it just happens to be Rust here"?

Because if not then I don't see why this does not leverage the type info of the request.
I see that `.send_json` takes `impl Serialize`, so passing a `BrpRequest` here would work.

Looking it up I see that the fields within `BrpRequest` are just `serde_json::Value` types right now so constructing a req is a bit awkward.
Maybe clients aren't expected to have any type info to lean on here, curious what others think?

File: crates/bevy_remote/src/lib.rs
Comment (mweatherley) [reply crates/bevy_remote/src/lib.rs:336]: Now that I look at the code again, it looks like we also didn't update this to actually check the incoming `jsonrpc` version number for requests; I had just always been testing with `"jsonrpc": "2.0"` which was swiftly ignored. I guess we didn't catch this earlier because of how `serde` deserializes json maps into structs.

Comment (torsteingrindvik) [reply crates/bevy_remote/src/lib.rs:336]: In addition to updating the docs I belive the `BrpRequest` struct itself should have this field

```rust
    /// This field is mandatory and must be set to `"2.0"`.
    pub jsonrpc: &'static str,
```

copied from `BrpResponse`

Comment (torsteingrindvik) [reply crates/bevy_remote/src/lib.rs:336]: Looking at the compatibility section of the spec:

> JSON-RPC 2.0 Request objects and Response objects may not work with existing JSON-RPC 1.0 clients or servers. However, it is easy to distinguish between the two versions as 2.0 always has a member named "jsonrpc" with a String value of "2.0" whereas 1.0 does not. Most 2.0 implementations should consider trying to handle 1.0 objects

I suppose not caring if clients put the `2.0` thing in requests is a way of handling v1 clients, if that's the case then maybe leaving it out of the struct isn't a bad idea.

File: examples/remote/client.rs
Comment (alice-i-cecile) [reply examples/remote/client.rs:44]: An assortment of opinions for you:

1. Clients can be written in any language.
2. Rust clients should be assumed to be aware of the `brp` types, but not any component / resource types.
3. We should demonstrate how to best do this in Rust code in our examples.

Comment (mweatherley) [reply examples/remote/client.rs:44]: At least as currently designed, `BrpRequest` is not really the ideal thing for the client side, since it delays deserializing the fields for the sake of error reporting (and doesn't restructure them back into a "finished thing" before processing the result of that transformation). Honestly, I'm not sure exactly how valuable or load-bearing that choice is; it's probably worth tinkering with.

On the other hand, the types like `BrpGetParams` in `builtin_methods` could plausibly be used to make constructing the parameters for the requests more straightforward, although they were not really designed with that in mind. (In general, the Rust types are intended primarily for use on the server side.)

Comment (mweatherley) [reply examples/remote/client.rs:44]: Okay, I tinkered around with this and reached the conclusion that there isn't actually any reason to defer parsing on the `method` field, but the other fields are still basically required to be arbitrary optional JSON values in order to support the JSON-RPC spec. With some tweaks to the serialization derive, that should also be roughly suitable for client-side use (at least in the `client` example).

Comment (mweatherley) [reply examples/remote/client.rs:44]: Taking *another* another look at this, it seems that maybe the reason it is structured this way is so that the `id` for the request can be reported if it is successfully parsed, even if other fields are missing and/or incorrect. I think that's actually fairly valuable, so I'm going to stick with that philosophy for now. Maybe in the future, it would be possible to restructure these in a way that makes the server and client side data types line up nicely.

Comment (ChristopherBiscardi) [reply examples/remote/client.rs:44]: When I changed this my only intention was to do something more reasonable than [manually constructing a string of json](https://github.com/bevyengine/bevy/pull/14880/files/ac96b055e8bb97b588d1ec221d6c29794b382e60#diff-b9891fe288692251ee4e78ed3d86ebbd13b187554d075ceb9f7e8d067d5fcce7R67-R76) so no strong attachment to this implementation from that perspective.

**Review by torsteingrindvik — APPROVED** [review]

**Comment by mweatherley** [comment]
All right, in light of @torsteingrindvik's comments, I lightly restructured `BrpRequest` so that it doesn't need to parse parameters lazily any more. This means it can be used more easily to structure input for clients, which I have demonstrated in the `client` example (which doesn't need raw JSON any more). The rough edges are still there, though; if we want to support this kind of thing more thoroughly, we should actually add useful constructors for these types. Obviously, I consider that out of scope for the present PR. :)

File: crates/bevy_remote/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_remote/src/lib.rs:449]: Why does this function take `anyhow::Error` while `BrpError::internal` takes `impl ToString`? Should we have both take `impl ToString` since the `anyhow::Error` is immediately converted to a string?

File: crates/bevy_remote/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_remote/src/lib.rs:479]: Question: For unfamiliar contributors[^1], would it make sense to add a comment explaining how these integers were chosen?

[^1]: me lol

File: crates/bevy_remote/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_remote/src/lib.rs:578]: Nit: Any reason to not do `name: impl ToString` or `name: impl Into<String>`? I see `RemoteMethods::insert` at least uses the latter.

File: crates/bevy_remote/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_remote/src/lib.rs:549]: Nit: Why are some impls next to their types and others away from them? The macro crate for `bevy_reflect` does a similar thing (which I think was my fault haha) and it can be slightly annoying jumping around between definition and implementation.

File: crates/bevy_remote/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_remote/src/lib.rs:593]: Nit: Should these be constants?

File: crates/bevy_remote/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_remote/src/lib.rs:631]: Nit: Should we document anywhere that a `RemotePlugin` instance cannot be built to multiple `App` instances (due to this `drain`)?

File: crates/bevy_remote/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_remote/src/lib.rs:672]: What happens if someone mutates the `HostPort` or `HostAddress` resources? Do we not have a way to kill this task if we want to restart the server on a new port/address?

File: crates/bevy_remote/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_remote/src/lib.rs:783]: Nit: You could also do
```suggestion
        Ok(BrpBatch::Single(request)) => {
            serde_json::to_string(&process_single_request(request, request_sender).await?)?
        }
        Ok(BrpBatch::Batch(requests)) => {
            let mut responses = Vec::new();

            for request in requests {
                responses.push(process_single_request(request, request_sender).await?);
            }

            serde_json::to_string(&responses)?
        }
```

File: crates/bevy_remote/src/builtin_methods.rs
Comment (MrGVSV) [inline crates/bevy_remote/src/builtin_methods.rs:35]: By "full paths", are we talking `std::any::type_name` or `TypePath`? I think wherever we mention "type path", "full path", etc. we should have that link to the actual one being used.

**Review by MrGVSV — APPROVED** [review]
I know I'm super late to reviewing this, so feel free to punt/disregard my comments 😅

File: crates/bevy_remote/src/lib.rs
Comment (mweatherley) [reply crates/bevy_remote/src/lib.rs:160]: If there is a `result` at all, that means that the request succeeded. If there is an `error`, it failed. It's basically like `Result<T, E>` in that way, if you want to think of it like that.

Comment (MrGVSV) [reply crates/bevy_remote/src/lib.rs:160]: Right but I’m thinking of someone writing something like a JavaScript application to interface with this. They would have to use `response.hasOwnProperty("result")`, `response.error === undefined`, or `response.result !== undefined`. And it would be an easy mistake to use `response.error !== undefined` or `response.result !== null`, since those will only return `true` on success for any non-null-returning request.

I guess this shouldn't be an issue if they read the documentation and understand what kinds of responses to expect, but something to consider.

Comment (mweatherley) [reply crates/bevy_remote/src/lib.rs:160]: I see what you mean, but this is also just how success and error are defined in response objects according to the JSON-RPC spec, so they would have to do that in order to handle responses correctly in any case.

File: crates/bevy_remote/src/lib.rs
Comment (mweatherley) [reply crates/bevy_remote/src/lib.rs:479]: I added a comment about the range reserved by the JSON-RPC specification; outside of that, the Bevy errors numbers are chosen more-or-less arbitrarily.

File: crates/bevy_remote/src/lib.rs
Comment (mweatherley) [reply crates/bevy_remote/src/lib.rs:578]: Good idea!

File: crates/bevy_remote/src/lib.rs
Comment (mweatherley) [reply crates/bevy_remote/src/lib.rs:549]: (Because this was basically co-written by three people :P ) That also bothers me, so I'll try to change it.

File: crates/bevy_remote/src/lib.rs
Comment (mweatherley) [reply crates/bevy_remote/src/lib.rs:593]: Yeah, they probably should.

File: crates/bevy_remote/src/lib.rs
Comment (mweatherley) [reply crates/bevy_remote/src/lib.rs:672]: Nope! At the moment the intention is that these resources are populated once with the data from the `Plugin` and are never supposed to be changed. Making it so the server can be restarted might be nice, though.

File: crates/bevy_remote/src/lib.rs
Comment (MrGVSV) [reply crates/bevy_remote/src/lib.rs:160]: Ah okay that makes sense it's part of the spec. Thanks for the clarification!

File: crates/bevy_remote/src/lib.rs
Comment (MrGVSV) [reply crates/bevy_remote/src/lib.rs:672]: Gotcha! Maybe until then we add doc comments on those resources saying that changing them currently does nothing and that they shouldn't be modified?

File: crates/bevy_remote/src/lib.rs
Comment (mweatherley) [reply crates/bevy_remote/src/lib.rs]: I don't think the plugin can be added by reference anyway?

Comment (MrGVSV) [reply crates/bevy_remote/src/lib.rs]: Can't you do `plugin.build(&mut app);`?

Comment (alice-i-cecile) [reply crates/bevy_remote/src/lib.rs]: Man if someone is doing that they deserve the bugs >.>

**Comment by alice-i-cecile** [comment]
Of course, wasm build failure... Ping me when you want me to retry!

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1696 if you'd like to help out.

**Pull request merged**
